### ISSUE TITLE
QUESTIONS: Changing Outputs for TTS & Averaging Checkpoints

### ISSUE THREAD

**Issue by Imtinan1996** [espnet/espnet#1424]
I am using Tacotron2 trained on a custom dataset on V3 Config, along with a custom trained ParallelWaveGAN FROM @kan-bayashi implementation. Everything works well but i have two concerns.

Q1 - Is there a better way to average checkpoints? for TTS checkpoint averaging is done over the last n files, whereas for ASR it looks for val/mean/acc in the train log, and averages the best n files. Can something like this not be achieved in TTS? I understand we don't have accuracy per say, but i tried averaging over the checkpoints with the least "validation/main/loss" will that be fine?

Q2 - I have realized that upon testing the exact same input multiple times, i get slightly different outputs in terms of the tone of the speaker. This doesn't corrupt the intelligibility of the speaker. But technically isn't the model supposed to be "frozen" for inference. Shouldn't the exact same input give the exact same output every time? or is there a random seed that causes these small fluctuations? I fear that the model is not frozen for inference and somehow some weight updation is going on. Maybe this will cause problems if a model is used for too long.

Also its worth to be noted that i do not use tts_decode.py , rather i have created a FLASK server that "forever" hosts the model, this is why i noticed such changes. Running synth_wav.sh  multiple times with the same input always gives the same output, but by giving the same input multiple times to synth_wav.sh in the same batch, thats when we get different generations

![image](https://user-images.githubusercontent.com/30050508/70501858-afd03d80-1b38-11ea-8bd5-3eadc4950f43.png)
Here is a comparison of two different wavs generated via the same input

**Comment by kan-bayashi** [comment]
Hi @Imtinan1996.

> Q1 - Is there a better way to average checkpoints?

Actually, I did not strictly check the effectiveness of the checkpoint averaging.
The difficult point is that while in the case of ASR, the validation accuracy is directly correlated to the WER or CER, the validation loss is not directly related to perceptual quality in the case of TTS. In my feeling, the validation loss best model is worse than the last epoch model but I have no idea when we should stop the training. And note that the checkpoint averaging is effective for Transformer but not so much for RNN-based models in the case of ASR. 

Do you have any idea? @r9y9 

> Q2 I have realized that upon testing the exact same input multiple times, i get slightly different outputs in terms of the tone of the speaker. 

This is because the use of dropout in the inference. 
https://github.com/espnet/espnet/blob/3a4b65e95289fabbf6500d75336a60085e61c42e/espnet/nets/pytorch_backend/tacotron2/decoder.py#L97-L102
I followed the manner of the original Tacotron 2 paper. If you want to fix the outputs, you can try to disable dropout in the inference. But I'm not sure the quality is changed or not when we remove dropout in the inference.

>  i have created a FLASK server that "forever" hosts the model

It seems nice :) I also want to make such script to be used as TTS API.

**Comment by Imtinan1996** [comment]
Thank you for your response @kan-bayashi 

> I have no idea when we should stop the training.

I have faced similar problems. In my case the model at epoch ~800 performs better than at epoch ~1000. I guess for now its just manual opinion based evaluation.

> This is because the use of dropout in the inference.

Oh i missed that note, i'll try changing the dropout to see if it brings any changes into the output. 

> It seems nice :) I also want to make such script to be used as TTS API.

I actually have written a script, this also hosts a html frontend for easy inference, i would love to contribute to the API by sharing this if you'd like me to.

**Comment by r9y9** [comment]
I don't have enough experience on the model averaging and I have no idea when to stop training :( As @kan-bayashi said, validation losses are not necessarily related to perceptual quality and it is difficult to choose the best model based on the validation values. That being said, I think it's worth monitoring CER/WER during training and it would be helpful for model selection. In that case, we need an external ASR model and may complicate the implementation. I was also thinking that we could have [MOSNet](https://arxiv.org/abs/1904.08352) or similar to estimate the subjective quality, but there are no action plans at the moment.

As for the second question, that's the intentional behavior as the original Tacotron paper suggests. Switching off the dropout at the inference may lose the robustness of the model. Discussions can be found in https://github.com/mozilla/TTS/issues/50. See also sec. 2.3 of https://arxiv.org/pdf/1811.07240.pdf.

**Comment by unilight** [comment]
I can provide some personal opinions regarding model selection in TTS.

As mentioned above, the validation loss does not correspond to perceptual quality. (I guess this is because validation loss is calculated under teacher forcing?) So is the case for testing time distortion measures. (MSE, MAE, MCD, etc.) As for WER/CER, I think it is good if you want to monitor _intelligibility_ but again,  it does not completely correspond to naturalness.

In my experience, as the training goes, the model will first generate very mono-toned, robot-like voice. Then, fine details such as prosody and intonation will be gradually generated. However, good WER/CER can be obtained even with the robot-like voice. I suspect that, all the ASR model ask is a _clearly_ uttered sentence. Whether it is _naturally_ uttered or not is not its main concern. One evidence to support this hypothesis is that, I've tried to measure WER/CER of the same test set with GL and Parallel WaveGAN. Surprisingly, even with the unnatural voice generated by GL, they actually give similar WER/CER.

One last remark is that, though not yet rigorously verified, it seems that with sufficient amount of data, E2E TTS models like Tacotron or Transformer-TTS do not overfit. So, it might be better to just train for a long time.
